### ISSUE TITLE
Option DLIBICAL_SYNCMODE_THREADLOCAL doesn't account for memory leaks

### ISSUE THREAD

**Issue by vtronko** [libical/libical#662]
I spotted a problem which was hard to explain. I had a rather simple program that parses libical file, and when I tested it with `-fsanitize=address`, it reported huge memory leaks.

I couldn't reproduce it with a minimal reproducible example program, but then I realized that it matters to run the parser not in the main thread, and was able to reproduce:
`SUMMARY: AddressSanitizer: 80661 byte(s) leaked in 421 allocation(s).`

Test program here:

```
#include <fstream>
#include <future>
#include <iostream>
#include <sstream>

#include <libical/ical.h>
#include <libical/icalparameter_cxx.h>
#include <libical/icalproperty_cxx.h>
#include <libical/icalvalue_cxx.h>
#include <libical/vcomponent_cxx.h>

int main() {
  std::string datastr;
  std::ifstream myfile("/tmp/test.ics");
  std::stringstream ss;
  ss << myfile.rdbuf();
  datastr = ss.str();


  auto f = std::async(std::launch::async, [&] {
    LibICal::VComponent comp(
        icalparser_parse_string((const char *)datastr.c_str()));
  });

  f.wait();

  return 0;
}
```

Compiling test case with:
`clang++ /tmp/test.cc -std=c++20 -g3 -O0 -fno-strict-aliasing -lical -lical_cxx -licui18n -licuuc -fsanitize=address`

Flags I use to build my local libical are here:
```
cmake -B build -G Ninja -DSHARED_ONLY=OFF -DICAL_BUILD_DOCS=OFF -DSTATIC_ONLY=ON -DCMAKE_INSTALL_PREFIX=libical/usr/local/ -DCMAKE_INSTALL_LIBEXECDIR=lib -DGOBJECT_INTROSPECTION=false -DENABLE_GTK_DOC=OFF -DCMAKE_BUILD_TYPE=RelWithDebInfo -DCMAKE_EXPORT_COMPILE_COMMANDS=ON -DLIBICAL_SYNCMODE_THREADLOCAL=true -DLIBICAL_DEVMODE_MEMORY_CONSISTENCY=OFF -DICAL_GLIB=False -DCMAKE_VERBOSE_MAKEFILE:BOOL=ON
```

Simply changing `s/launch::async/launch::deferred/` makes problem go away.
    
------

According to https://github.com/libical/libical/pull/609, the main point was to allow memory safety in parallel contexts.
Tests in `regression.c` don't actually try to check how `icalparser_parse_string` works in parallel threads.

So I picked a random test that used the `icalparser_parse_string`, and applied following diff (pardon my C):
```diff
diff --git a/src/test/regression.c b/src/test/regression.c
index 3584436b..372d4281 100644
--- a/src/test/regression.c
+++ b/src/test/regression.c
@@ -699,6 +699,14 @@ void test_component_foreach(void)
     }
 }

+void* pthreadWrapper(void* u) {
+  (void)(u);
+    // Call your C function with no arguments
+    test_component_foreach();
+
+    return NULL;
+}
+
 void test_recur_iterator_set_start(void)
 {
     icaltimetype start = icaltime_from_string("20150526");
@@ -5463,6 +5471,13 @@ int main(int argc, char *argv[])
     test_run("Test Properties", test_properties, do_test, do_header);
     test_run("Test Components", test_components, do_test, do_header);
     test_run("Test icalcomponent_foreach_recurrence", test_component_foreach, do_test, do_header);
+
+    pthread_t thread;
+    if (pthread_create(&thread, NULL, pthreadWrapper, NULL) != 0) {
+      return 1;
+    }
+    pthread_join(thread, NULL);
+
     test_run("Test icalrecur_iterator_set_start with date", test_recur_iterator_set_start, do_test, do_header);
     test_run("Test weekly icalrecur_iterator on January 1", test_recur_iterator_on_jan_1, do_test, do_header);
     test_run("Test Convenience", test_convenience, do_test, do_header);
```

And regression test started to report leaks:

```
SUMMARY: AddressSanitizer: 108269 byte(s) leaked in 460 allocation(s).
```

Which would be one thing if it's capped at some point, and just freed upon process exit, but I easily reproduced that this can leak gigabytes of memory in runtime, so it's kinda worrisome.

**Comment by minichma** [comment]
The overhead per thread is indeed the trade-off of using `-DLIBICAL_SYNCMODE_THREADLOCAL=true`. By using it, all global variables (most notably the the time zone cache), are allocated per thread rather than globally. This allows for lock-free programming but has some caveats. One is that you need to think about the max number of threads and the overhead associated with it. To work around the issue you have several options:

* Not using `-DLIBICAL_SYNCMODE_THREADLOCAL=true`: This would be the classic, most straight-forward way. If pthreads is available, the build would fall back to pthread-based syncing. Globals would be allocated just once. However, if you plan to use libical in multiple threads concurrently, you'd have to apply additional synchronization in order to avoid race conditions, e.g. related to the error state (you can look up all related global variables by searching for `ICAL_GLOBAL_VAR` in the libical source base).
* Using libical from just a single thread.
* Release allocated memory after calls into libical on each thread, most notably by calling `icaltimezone_free_builtin_timezones()`. However, be aware of the overhead involved in populating the cache of built-in time zones on each call!
* Make sure, threads are removed after individual calls to libical. When using std::launch, introduce some scheduler that limits the overall number of threads to a reasonable number or that spawns a new thread before and removes it after each call. In this case you also need to be a ware of the overhead of populating the built-in time zones cache on each call!

A nice option would be to have just a single global cache of built-in timezones (which should be considered immutable), while having other global variables marked thread-local. However, that's not something that exists as of today.

**Comment by winterz** [comment]
This is great info that we need to add to our documentation.  Maybe it fits
onto the wiki somewhere at https://github.com/libical/libical/wiki
If someone wants to do that -- or maybe even better in doc/UsingLibical.md
in the 5.5 Memory Management Section (or somesuch)

**Comment by minichma** [comment]
Agree. I can work on that as soon as I find some time, maybe next week.

**Comment by vtronko** [comment]
My bad, overlooked that I pasted wrong option name in title, but I assume it was clear from the ticket description.

**Comment by minichma** [comment]
Added some explanation to the CMAKE param in #672. However, a more comprehensive documentation should be added to `UsingLibical.md`, but this is out of scope of what I can do any time soon, as documentation on multithreading seems to be missing altogether. We should add a section explaining things like:
 * Which kind of data could be shared among threads? E.g. library configuration (see `icalerror_set_errors_are_fatal()`, `icalmemory_set_mem_alloc_funcs()`,
  `icaltimezone_set_builtin_tzdata()`, `ical_set_unknown_token_handling_setting()`, `set_zoneinfopath()`, etc.), 
 * pthreads
   * When is it used?
   * What does it do, which level of synchronization does it provide? E.g. the last error code as well as the ring-buffer are already stored/allocated thread-locally, resulting in the requirement, e.g. to access the error code from the same thread that produced it. Access to the cached zone data is synchronized.
 * thread-local
   * still marked experimental
   * overhead regarding memory and initialization time (mainly due to the built-in time zones).

Regarding the built-in timezones it could make sense for the thread-local mode to still allocate it only once, but this would require additional synchronization for that part.

@winterz How should we proceed on that? Create additional ticket(s) or use this one?

**Comment by winterz** [comment]
please make a new issue and close this one.
you can copy+paste your very nice outline in the previous comment to the new issue.

**Comment by minichma** [comment]
Closing this as the options works as expected. Added some description to the CMAKE option in #672 and created a follow-up ticket for additional documentation in #673. Also note that the whole option is still marked experimental.

**Comment by vtronko** [comment]
Thanks for update, @minichma.
So do I understand correctly, that meanwhile I should call `icaltimezone_free_builtin_timezones` in between the calls? Should I do it after every libical call, or does it suffice to call it once I am cleaning the component after I am done with parsing document?

**Comment by minichma** [comment]
First I suggest to double-check whether you really need to use the thread-local model. While it offers straight-forward synchronization, it's still marked experimental, so it might not be the best choice for production scenarios. If you are able to set the configuration (icalerror_set_errors_are_fatal(), icalmemory_set_mem_alloc_funcs(),
icaltimezone_set_builtin_tzdata(), ical_set_unknown_token_handling_setting(), set_zoneinfopath(), etc.) globally rather than per thread, then the pthreads model would be the better choice for you. The pthread model would be used by setting `-DLIBICAL_SYNCMODE_THREADLOCAL=false` (assuming that pthreads is available on your platform).

If you want to continue using the thread-local model, consider what I wrote in my initial response
> * Release allocated memory after calls into libical on each thread, most notably by calling `icaltimezone_free_builtin_timezones()` [edit] and `icalmemory_free_ring()`. However, be aware of the overhead involved in populating the cache of built-in time zones on each call!
> * Make sure, threads are removed after individual calls to libical. When using std::launch, introduce some scheduler that limits the overall number of threads to a reasonable number or that spawns a new thread before and removes it after each call. In this case you also need to be a ware of the overhead of populating the built-in time zones cache on each call!

However, I didn't try this. There might be other memory that needs to be freed and some static primitives will still stay around. So it would also depend on your memory checker on whether it would still complain.

**Comment by minichma** [comment]
Sorry for being unclear. The functions I mentioned (`icalerror_set_errors_are_fatal()` et al), and some others are the ones NOT being synchronized. If you don't use them, you probably don't have any issues with synchronization and you can use the pthreads synchronization model. *If* you use them, but only set them once at startup time rather than in each thread individually, then you are also fine with the pthreads model. However, *if* you use them in each thread individually, possibly with different values per thread, then you need to take care of synchronization in some way, but this case is rather rare. In this case the thread-local model would be helpful, in all other cases you're probably fine with pthreads.

> Also, why close the issue?

Because the ticket describes a problem that doesn't exist in this way. If you want to create a follow-up ticket and maybe contribute a PR, I'm sure, you're welcome to do so. However, completing the thread-local mode would require some conceptual work, most notably for the way of how to deal with the timezone cache. Should it be kept per thread or cached just once? Probably just once but then we need to think about syncing again, so we wouldn't reach a fully sync-free solution. Besides this mainly the tests would be missing, you'd certainly be welcome to contribute those.
